### ISSUE TITLE
wsl leaves multiple background hosts alive

### ISSUE THREAD

**Issue by leongrdic** [Microsoft/WSL#3503]
* Your Windows build number:  Microsoft Windows [Version 10.0.17134.228]

* What you're doing and what's happening:

1. i started a powershell window and typed `wsl`
2. in wsl i started an `ssh-agent`
3. i typed exit and closed the powershell window
4. repeat the previous steps a few time

we can now see multiple powershell hosts with their child processes running in the background and we can see all the `ssh-agent` processes in `top` in a new powershell window.

![image](https://user-images.githubusercontent.com/6260068/44612836-2e9e4b80-a80c-11e8-97dc-d29ae82b19ea.png)

* What's wrong / what should be happening instead:

when i manually end any number of background powershell processes, all the `ssh-agent`s are still alive **as long as i got at least one powershell (background or foreground) task running**, so my conclusion is only one background powershell (or console host or whatever) task is needed for keeping all wsl processes alive. when i end the last powershell task through task manager, the whole wsl suddenly stops and all wsl processes get terminated.

based on this behavior, there's no point in keeping a number of powershell tasks open in background. this way I found myself having a few dozens of background powershells open while regularly using wsl.

this problem doesn't only apply to `ssh-agent`, the same thing happens with any process inside wsl that starts in the background.

**Comment by therealkenc** [comment]
> my conclusion is only one background powershell (or console host or whatever) task is needed for keeping all wsl processes alive.

Zero (not one) powershell process(es) are needed to keep wsl processes "alive". The analogy is if you `telnet`ed into a Linux cloud server, started `ssh-agent`, exited the `telnet` session, and then exited PS. Another analogy would be if you started a win32 ssh-agent service, and exited PS. If you want `ssh-agent` dead, you have to stop it gracefully or kill it with fire. Blog post is [here](https://blogs.msdn.microsoft.com/commandline/2017/12/04/background-task-support-in-wsl/).

**Comment by leongrdic** [comment]
> Zero (not one) powershell process(es) are needed to keep wsl processes "alive"

I agree that this should be the expected behavior but when I end the last powershell (background) task, the wsl suddenly stops and all proccesses I started inside of wsl get terminated. 

I read the blog post before opening the issue.

Are you sure this is by design, as this way users could potentially open hundreds of background powershell (or cmd) tasks after a few days/weeks

**Comment by leongrdic** [comment]
> I can't reproduce this here. If I start one (and only one) powershell process, start wsl.exe, execute sudo service start cron, exit the WSL bash shell, and exit powershell, then cron remains running.

while trying to reproduce this, have you noticed that after exiting powershell (typing `exit` or clicking the X button on the window), it still stays as a background task (you can see it in task manager)?

if you then opened another powershell, run wsl and started another background service (such as `sshd` or `nginx` or `ssh-agent`), and then proceeded to exit that powershell window, it would also go to the background tasks section of your task manager leaving you with two powershell background tasks running.

now if you tried ending one powershell background task (through the task manager), both of your wsl background processes would continue working, no matter which powershell background task you ended. however if you end the second (last) powershell task, then the whole wsl stops and all your background services get terminated.

what i've been trying to say is, there's no need for keeping all the powershell tasks in background.

**Comment by leongrdic** [comment]
Thanks for your reply and patience. I'll try to find some time to setup a vm running the insider build you're using and will try to reproduce this behavior.

**Comment by therealkenc** [comment]
> ...I can't reproduce your accumulating PowerShell instances ... I have no clear explanation to give you on that. 

Okay I have a working theory, though I'm a little sleep deprived and could be wrong. I think the change in the observed behavior _might_ be because of the [ConPTY changes](https://blogs.msdn.microsoft.com/commandline/2018/08/02/windows-command-line-introducing-the-windows-pseudo-console-conpty/), with a nod to #2370 ([message](https://github.com/Microsoft/WSL/issues/2370#issuecomment-372051463)) which shipped in [17618](https://docs.microsoft.com/en-us/windows/wsl/release-notes#build-17618-skip-ahead). I don't want to speculate too much, but at base, I couldn't see how your "background" PowerShells l could have to do with WSL proper. Those PowerShell instances aren't any of WSL's business (as alluded previously, you could have come in through say PuTTY and started `nginx`, and WSL sure as **** isn't going to hold PuTTY hostage in the background once you exit). I almost said out loud in the previous post "maybe something in PowerShell changed", but that didn't make much sense either. But something _did_ change in the way PTYs are done.

Recommendation would be to try Insiders and see if your problem "goes away". If my theory is correct this was a long meandering non-obvious walk to dupe.

**Comment by DHowett** [comment]
ConPTY is currently only used when hosting Windows processes from WSL, not the other way around.

**Comment by therealkenc** [comment]
Yeah I know. I didn't want to get in the weeds ("speculate too much"). Idea was that before the improvements, interop needed to hold a hidden console hostage, despite the TTY being detached by the daemon. `LxManager.dll` would then refcount them, and clean up the turds when necessary. I still could be very wrong. I wouldn't have said anything except late and weekend. Beats "closed - works for me".

**Comment by therealkenc** [comment]
Alright. I'm downloading a 17713 to look. In the mean time, what are your CLI steps just to be sure I do the exact same repro. Don't need screenshots. Just lines of text that can be cut and paste into two PS consoles before they're closed. If this reproduces in 17713 but not 18219 this is going to get a big grumble (which isn't on you).

**Comment by leongrdic** [comment]
alright, so start up a new powershell and type:
```
wsl
sudo service ssh start
exit
exit
```

for the second powershell you repeat the same steps but instead of ssh start any other background service e.g. cron, nginx, etc. or start an `ssh-agent`:

```
wsl
sudo service cron start
exit
exit
```

take a look at your task manager and you will see two background powershells running

> If this reproduces in 17713 but not 18219 this is going to get a big grumble

why is that if i may ask?

**Comment by therealkenc** [comment]
> why is that if i may ask?

Because it's a lot of effort just to find out it was fixed in skip-ahead. Makes for a questionable Sunday time-sink.

**Comment by leongrdic** [comment]
oh ok get it. im sorry if thats the case
speaking of, i tried upgrading to 18219 through regular windows update but its throwing me error `0xc0000005`, tried multiple times after restarts and switching to fast in insider settings. i dont know how else to get the newest build since i cant find download iso anywhere

**Comment by therealkenc** [comment]
This is on a shiny new 17713 in a VM. Below is my background task list in Task Manager, with `sshd` running but no "Windows Powershell (n)" or "Windows Command Processor (n)" in the Double-ues. `cron` is in the scroll-up. Dunno what to tell you at this point.

![image](https://user-images.githubusercontent.com/7049700/44632977-2859bd80-a939-11e8-8026-e6170a4818ce.png)

**Comment by leongrdic** [comment]
how is that possible... im really confused, my installation is also fresh..
which distro did you install?

**Comment by therealkenc** [comment]
Ubuntu 18.04 from the MSFT Store. First thing I did after enabling WSL in 'features'. Thing I did after that was your repro steps verbatim. I didn't even stop for a coffee.

**Comment by leongrdic** [comment]
huh. i could try to test this on another computer, i have the 17134 build on my physical laptop and i emulated the 17713 build in virtualbox on the same laptop. both are x64. i doubt that could cause such a problem but i guess its worth testing.

**Comment by therealkenc** [comment]
> but i guess its worth testing

[_Effort_](https://www.urbandictionary.com/define.php?term=effort). Me I'd sit and wait for a "I'm seeing this too" from someone else, and maybe an explanation from the devs. I mean, okay, you try again on another box and find (a) same thing or (b) no PowerShells held hostage like me. Either way, then what. Suppose it depends a lot on how hard up you are for a time sink today.

**Comment by leongrdic** [comment]
hahah okay i see your point.
could you somehow ping the devs to read the issue and try to explain this behavior?

**Comment by ThisInsider** [comment]
![image](https://user-images.githubusercontent.com/17685101/44634137-bf546300-a995-11e8-94a8-342fd43dbdcc.png)
I can confirm this behavior. Running windows 10 x64 build 17134

**Comment by leongrdic** [comment]
you're right. this is very interesting. i didn't even think about checking process explorer or the details tab.
though, i think my original question still remains - but for `wslhost.exe` instead of powershell.
why do multiple `wslhost.exe` processes stay running in background? it's unnecessary and resource consuming (probably insignificant, but why keep e.g. 10 instances in background if a single one is enough)?

**Comment by Biswa96** [comment]
> why do multiple wslhost.exe processes stay running in background? 

Probable cause: If one background WSL-ed Linux service (sshd etc.) fails/corrupts it'll not affect other services.

**Comment by leongrdic** [comment]
that would make sense if a new `wslhost.exe` was forked for each new wsl service, but that's not the case. even if you start multiple background services or processes in a wsl session, only a single `wslhost.exe` will be spawned (along any old `wslhost.exe`s)

also, out of curiosity, i tried killing the `conhost.exe` process that is left as a child process of **each** `wslhost.exe` and it doesn't effect wsl and its services.

so basically a single `wslhost.exe` is needed for the whole wsl to work in background.

**Comment by therealkenc** [comment]
> i think my original question still remains - but for wslhost.exe instead of powershell

Nice to have an answer to that, given Dustin's excellent before and after pics. But at base the answer is "because the devs decided to do it like that" and we're basically back to by-design or question tag again.

[Better question would be why the Rube Goldberg host service on top of host service architecture when there was no backward compatibility constraints. A `devpts.sys` filesystem [could be](https://github.com/Microsoft/WSL/issues/2370#issuecomment-372078145) mounted in Windows, meaning [porting](https://github.com/Microsoft/vcpkg/tree/master/ports) PTY code would amount to changing the path string instead of having to shim a whole new obstinately different API. But I digress.]

> i tried killing the conhost.exe process .., and it doesn't effect wsl

No.

**Comment by piotrpalek** [comment]
I have an issue (I'm on the latest windows update) that's very similar. I've also followed this guide: https://evalcode.com/zsh-default-wsl-windows-10/ to setup ZSH on WSL (not sure if relevant)

Now I've started a shell, started a tmux server and attached to it. And now every time I close the terminal, reopen it and attach to the existing tmux server a new ZSH process keeps running in the background. That doesn't sound right and surely isn't how it works on linux (or I just didn't notice, correct me if I'm wrong).

**Comment by benhillis** [comment]
@leongrdic - Wslhost is used so background processes (sshd in your example) are able to use NT interop (launching Windows binaries) with the right permission / environment as the process that launched WSL. The overhead of the process is minimal and the binary only does "work" when interop requests are received.
